include/zephyr/spinlock.h

```diff
@@ -153,6 +153,9 @@ static ALWAYS_INLINE k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
 
 #ifdef CONFIG_SMP
 	while (!atomic_cas(&l->locked, 0, 1)) {
+#ifdef CONFIG_ARCH_HAS_BUSY_SPINLOCK_CHECK
+		arch_busy_spinlock_check();
```

Comment:
Pedantic API naming: can we name this something like "arch_spin_relax()" instead?  A more common and less obscurely-Zephyr-specific use case for this sort of thing is idle power management and bus contention relaxation.  x86 spinlocks are best implemented on big/NUMA systems with MWAIT, etc...  Our default spinlock is naive and works great in practice, but almost every system has a "better way to do this".

The other advantage is that we can then document this in such a way that it can be used by other busy loops than just "k_spinlocks".

Reply:
Also: another very reasonable implementation choice would be to make this a function call and implement the default as a weak symbol, which avoids the need to mess with kconfig.  Almost by definition, we don't care about cycle-level performance optimization when we're doing nothing waiting for something to happen.